### PR TITLE
Lax-based consumption function

### PR DESCRIPTION
**Pull request by aphearin**
**Do not merge:** this PR is only intended to illustrate a working prototype for a lax-based implementation of the **_get_lagged_gas** function. The new function **_dmgas_dt_kern** agrees with **_get_lagged_gas** to working precision, so this provides starter code for an overhaul to the internals of our gas consumption function. However, there is still more development required and more experiments to do before moving forward.

The goal of a lax-based implementation is to improve the memory efficiency of DiffstarPop. The current implementation of DiffstarPop is based on vmapping the **_sfr_history_from_mah** function, which @alexalar and I are finding to be sufficiently heavy that we overflow the memory of our laptops for Ngals >~10k. This heaviness is exacerbated for downstream applications of DSPS, and so I am hoping that by replacing `jax.vmap` with `lax.scan` that we can avoid some of the memory overhead.

In our model, there is a double loop that required to compute the SFH. First, at fixed `t_obs`, there is an inner loop over all previous timesteps, `t_acc<t_obs`, in which we sum up contributions to the SFR from previously-accreted parcels of gas. Then there is a second loop over the full collection of `t_obs`. The way the current implementation works is to populate a matrix of consumption function kernels: the value of the SFR at `t_obs` is computed from a particular row, and each column within that row stores the value of the consumption function for different times `t_acc`; columns with `t_acc>t_obs` are zeroed out, and the SFR is computed as a weighted sum of `dMgas/dt(t_acc)` using the values in the columns for the weights, where the weighted sum is computed by collapsing the columnar dimension of this matrix via an axis summation. 

The code in this PR replaces the above algorithm with one in which at fixed `t_obs`, we instead use `lax.scan` to do the inner loop over all the previously accreted gas parcels, which should hopefully be more memory-efficient. The new unit-testing module `test_lax_gas.py` enforces that the new implementation agrees within 1e-4 with the original. However, the performance of the code in this PR is terrible because the outer loop over `t_obs` is in pure python. This was only for prototyping purposes, and I can think of several different ways that this starter code can be turned into an actual implementation. What we should do is explore variations like the ones sketched below, and importantly: we should do the performance benchmarking within the context of a DiffstarPop calculation to identify the implementation that performs best when calculating SFHs of galaxy populations:

1. The current prototype loops over successively longer sub-arrays of `dmhdt_table`. Instead of this, we could try looping over the full input `dmhdt_table`, but manually zero out the entries with `t_acc>t_obs`, possibly interpolating from the precomputed `dmhdt_table` in case this approach requires values at different times.
2. For each iteration `t` of the loop over `t_obs`, create a small array from `t_min <= t <= t_obs`, and compute `log_mah` and `dmhdt` on-the-fly on this small array, and then integrate over the small array.
3. For the outer loop, we could either use `lax.scan` over `t_obs`, or alternatively we could just use `vmap`. This use of `vmap` would still avoid populating and collapsing a matrix. We should try both and compare both performance and memory efficiency (again, using a benchmark based on DiffstarPop)
4. In the code brought in with the #4 PR, what we do is compute the lagged gas entirely separately from the SFH, and then just multiply the result by the MS efficiency within the otherwise-unchanged [_sfr_history_from_mah](https://github.com/ArgonneCPAC/diffstar/blob/4d15a5b2fd2faa86311c543a151fee73a14bd7f1/diffstar/stars.py#L398) kernel. But it may be faster to calculate the SFH within the `lax.scan` loop, so that our calls to the efficiency function are integrated within this new lax.scan-based approach.

### CODE REVIEW

**Comment by aphearin** [comment]
One additional experiment we should try is to use `lax.fori_loop` rather than `lax.scan`. The reason is that `lax.scan` actually accumulates an intermediate array along the way to doing its sum, but all we ever care about for this inner loop is the _final_ value of the lagged gas, and so it may be even more memory efficient to use `lax.fori_loop`.

**Comment by aphearin** [comment]
#16 makes this PR obsolete.
